### PR TITLE
P2p ping

### PR DESCRIPTION
**Pull request by y0sher**

### CODE REVIEW

File: ping/ping.go [hunk 1]
```diff
@@ -1,179 +1,124 @@
...
+		p.logger.Info("handle ping request")
+		req := &pb.Ping{}
+		if err := proto.Unmarshal(msg, req); err != nil {
+			return nil
+		}
 
-	return response, nil
-}
+		for _, f := range p.pingCallbacks {
+			if err := f(req); err != nil {
```
Comment (antonlerner) [inline ping/ping.go:68]: When a callback fails, should we not process al other callbacks? also note that this is done without paralelization and a callback might block execution of other cllbacks

File: p2p/dht/dht.go [hunk 2]
```diff
@@ -185,57 +226,90 @@ func filterFindNodeServers(nodes []node.Node, queried map[string]struct{}, alpha
...
-
 	timeout := time.NewTimer(LookupTimeout)
 Loop:
 	for {
 		select {
-		case res := <-results:
+		case qres := <-results:
+
+			// we mark active nodes
+			queried[qres.server.String()] = qres.res == nil
```
Comment (antonlerner) [inline p2p/dht/dht.go]: queried should be true if result is not nil
`queried[qres.server.String()] = qres.res != nil`

**Review by antonlerner — APPROVED** [review]
Please run with --race to make sure there are no races

File: ping/ping.go (hunk 1 above)
Comment (y0sher) [reply ping/ping.go:68]: its actually like this by design.
its a base for #586. we want to make some blocking checks before replying

Comment (y0sher) [reply ping/ping.go:68]: ill add a note

**Pull request merged**